staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -398,10 +408,59 @@ func PollImmediateUntil(interval time.Duration, condition ConditionFunc, stopCh
 		return nil
 	}
 	select {
-	case <-stopCh:
+	case <-ctx.Done():
 		return ErrWaitTimeout
 	default:
-		return PollUntil(interval, condition, stopCh)
+		return PollUntilWithContext(ctx, interval, condition)
+	}
+}
+
+// PollImmediateUntil tries a condition func until it returns true, an error or stopCh is closed.
+//
+// PollImmediateUntil runs the 'condition' before waiting for the interval.
+// 'condition' will always be invoked at least once.
+// DEPRECATED: Use the PollImmediateUntilWithContext instead
+func PollImmediateUntil(interval time.Duration, condition ConditionFunc, stopCh <-chan struct{}) error {
+	ctx, _ := contextForChannel(stopCh)
+	return PollImmediateUntilWithContext(ctx, interval, condition)
+}
+
+// WaitWithContextFunc creates a channel that receives an item every time a test
+// should be executed and is closed when the last test should be invoked.
+// The context passed to the wait function can be used to detect parent context deadline.
+type WaitWithContextFunc func(ctx context.Context) <-chan struct{}
+
+// WaitForWithContext continually checks 'fn' as driven by 'wait'.
+//
+// WaitForWithContext gets a channel from 'wait()'', and then invokes 'fn' once for every value
+// placed on the channel and once more when the channel is closed. If the channel is closed
+// and 'fn' returns false without error, WaitForWithContext returns ErrWaitTimeout.
+//
+// If 'fn' returns an error the loop ends and that error is returned. If
+// 'fn' returns true the loop ends and nil is returned.
+//
+// ErrWaitTimeout will be returned if the context is done without fn ever
+// returning true.
+func WaitForWithContext(ctx context.Context, wait WaitWithContextFunc, fn ConditionFunc) error {
+	stopCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	c := wait(stopCtx)
+	for {
+		select {
+		case _, open := <-c:
+			ok, err := fn()
+			if err != nil {
+				return err
+			}
+			if ok {
+				return nil
+			}
+			if !open {
+				return ErrWaitTimeout
+			}
+		case <-ctx.Done():
```

Comment:
> and then invokes 'fn' once for every value placed on the channel and once more when the channel is closed

If the parent context is closed, we might hit the second case and `fn` would never be called?

Reply:
if the parent context is closed, wouldn't that propagate to stopCtx and the wait() will get a chance to gracefully close? now sure I understand the question.

Reply:
Yes, if `ctx` is closed, it would propagate to `stopCtx` and wait() would close `c`. However, you don't know which case of your select will be chosen. If the first, `fn()` will get called at least once. Alternatively, if the second one is chosen `fn()` may never get called. Or am I missing something? :)